analytic_engine/src/instance/open.rs

```diff
@@ -319,46 +318,63 @@ impl ShardOpener {
+            match stage {
                 // Only do the wal recovery work in `RecoverTableData` state.
                 TableOpenStage::RecoverTableData(ctx) => {
-                    let table_data = ctx.table_data.clone();
-                    let read_ctx = ReadContext {
-                        batch_size: self.wal_replay_batch_size,
-                        ..Default::default()
-                    };
-
-                    let result = match Self::recover_single_table_data(
-                        &self.flusher,
-                        self.max_retry_flush_limit,
-                        self.wal_manager.as_ref(),
-                        table_data.clone(),
-                        self.wal_replay_batch_size,
-                        &read_ctx,
-                    )
-                    .await
-                    {
-                        Ok(()) => Ok((table_data, ctx.space.clone())),
-                        Err(e) => Err(e),
-                    };
-
-                    match result {
-                        Ok((table_data, space)) => {
-                            *state = TableOpenStage::Success(Some(SpaceAndTable::new(
-                                space, table_data,
-                            )));
-                        }
-                        Err(e) => *state = TableOpenStage::Failed(e),
-                    }
+                    replay_table_datas.push(ctx.table_data.clone());
                 }
                 // Table was found opened, or failed in meta recovery stage.
                 TableOpenStage::Failed(_) | TableOpenStage::Success(_) => {}
                 TableOpenStage::RecoverTableMeta(_) => {
                     return OpenTablesOfShard {
-                        msg: format!("unexpected table state:{state:?}"),
+                        msg: format!(
+                            "unexpected stage, stage:{stage:?}, table_id:{table_id}, shard_id:{}",
+                            self.shard_id
+                        ),
                     }
-                    .fail()
+                    .fail();
+                }
+            }
+        }
+
+        let replay_mode = match self.recover_mode {
+            RecoverMode::TableBased => ReplayMode::TableBased,
+            RecoverMode::ShardBased => ReplayMode::RegionBased,
+        };
+        let mut wal_replayer = WalReplayer::new(
+            &replay_table_datas,
+            self.shard_id,
+            self.wal_manager.clone(),
+            self.wal_replay_batch_size,
+            self.flusher.clone(),
+            self.max_retry_flush_limit,
+            replay_mode,
+        );
+        let mut table_results = wal_replayer.replay().await?;
+
+        // Process the replay results.
+        for table_data in replay_table_datas {
+            let table_id = table_data.id;
+            let stage = self.states.get_mut(&table_id).unwrap();
+            let table_result = table_results.remove(&table_id).unwrap();
```

Comment:
Will panic here?

Reply:
`states` and `table results` are generated from `replay_table_datas`, it is ensured to contains related table id, I will add some comment to describe this later.

Reply:
Some comments are added to explain.